pio/irq/main.c

```diff
@@ -0,0 +1,174 @@
+ * Typically you do not want to do work in ISRs because the main thread
+ * has more important things to do. Because of that, when we get the ISR
+ * I'm going to copy the state machine that fired the ISR into this
+ * variable.
+ *
+ * Variable is volatile so that it doesn't get cached in a CPU register
+ * in the main thread. Without this it's possible that you never see
+ * irq_flags get set even though the ISR is firing all the time.
+ *
+ * Of course, you can really do whatever you want in the ISR, it's up to you.
+ */
+volatile uint8_t irq_flags = 0;
+
+/**
+ * This function is called when the IRQ is fired by the state machine.
+ * @note See enable_pio_irqs for how to register this function to be called
+ */
+void simple_irq_handler() {
+    // Check the interrupt source and set the flag accordingly
+    // This method can handle multiple simultaneous interrupts from the PIO
+    // since it checks all the irq flags. The advantage to this is that the
+    // ISR would execute only once when simultaneous interrupts happen instead
+    // of being executed for each interrupt individually.
+    // For other applications, make sure to check the correct pio (pio0/pio1)
+    irq_flags = pio0_hw->irq;
+    // Clear the flags since we've saved them to check later.
+    hw_clear_bits(&pio0_hw->irq, irq_flags);
+
+    // alternatively you could use pio_interrupt_get(pio0, #) to check a specific
+    // irq flag.
+}
+
+/**
+ * Lets the pico know that we want it to notify us of the PIO ISRs.
+ * @note in simple_irq.pio we enable irq0. This tells the state machine
+ *       to send the ISRs to the core, we still need to tell the core
+ *       to send them to our program.
+ */
+void enable_pio_irqs() {
+    // Set the function that will be called when the PIO IRQ comes in.
+    irq_set_exclusive_handler(PIO0_IRQ_0, simple_irq_handler);
+
+    // Once that function is set, we can go ahead and allow the interrupts
+    // to come in. You want to set the function before enabling the interrupt
+    // just in case. The docs say if an IRQ comes in and there's no handler
+    // then it will work like a breakpoint, which seems bad.
+    irq_set_enabled(PIO0_IRQ_0, true);
+}
+
+/**
+ * Loads simple_irq pio program into PIO memory
+ */
+void load_pio_programs() {
+    PIO pio = pio0;
+
+    // Load the program into PIO memory
+    uint offset = pio_add_program(pio, &simple_irq_program);
+
+    // Tell each state machine to run the program.
+    // They are allowed to run the same program in memory.
+    simple_irq_program_init(pio, 0, offset);
+    simple_irq_program_init(pio, 1, offset);
+    simple_irq_program_init(pio, 2, offset);
+    simple_irq_program_init(pio, 3, offset);
+}
+
+/**
+ * Writes to the tx fifo of the given state machine.
+ * This will make the simple_irq program send an ISR to us!
+ */
+void trigger_irq(int sm) {
+    printf("Triggering ISR from state machine %d\n", sm);
```

Comment:
ISR -> IRQ here? There's also a few comments that still say "ISR", where perhaps "IRQ" would be more appropriate? (although there _are_ some comments where it might be appropriate to stick to "ISR"?)

In my understanding (which may still be wrong, as I've not done much low-level coding), IRQ = Interrupt ReQuest, i.e. the actual hardware signal, and ISR = Interrupt Service Routine i.e. the callback function you've registered to handle the IRQ.